File: shopify-clone/frontend/src/app/core/services/auth.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface LoginRequest { email: string; password: string; }
export interface RegisterRequest {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  role: 'CUSTOMER' | 'ARTISAN' | 'MERCHANT';
  brandName?: string;
  bio?: string;
  location?: string;
}
export interface AuthResponse {
  accessToken: string; refreshToken: string; tokenType: string;
  userId: number; email: string; firstName: string; lastName: string; role: string;
}

@Injectable({ providedIn: 'root' })
export class AuthService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/auth`;

  login(request: LoginRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.baseUrl}/login`, request).pipe(
      tap(res => this.saveTokens(res))
    );
  }

  register(request: RegisterRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.baseUrl}/register`, request).pipe(
      tap(res => this.saveTokens(res))
    );
  }

  logout(): void {
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('user');
  }

  getToken(): string | null {
    return localStorage.getItem('access_token');
  }

  isLoggedIn(): boolean {
    return !!this.getToken();
  }

  getUser(): AuthResponse | null {
    const user = localStorage.getItem('user');
    return user ? JSON.parse(user) : null;
  }

  private saveTokens(res: AuthResponse): void {
    localStorage.setItem('access_token', res.accessToken);
    localStorage.setItem('refresh_token', res.refreshToken);
    localStorage.setItem('user', JSON.stringify(res));
  }
}

File: shopify-clone/backend/src/main/java/com/shopifyclone/auth/AuthService.java
package com.shopifyclone.auth;

import com.shopifyclone.auth.dto.AuthResponse;
import com.shopifyclone.auth.dto.LoginRequest;
import com.shopifyclone.auth.dto.RegisterRequest;
import com.shopifyclone.domain.artisan.ArtisanProfile;
import com.shopifyclone.domain.merchant.MerchantProfile;
import com.shopifyclone.domain.user.Role;
import com.shopifyclone.domain.user.User;
import com.shopifyclone.repository.ArtisanProfileRepository;
import com.shopifyclone.repository.MerchantProfileRepository;
import com.shopifyclone.repository.UserRepository;
import com.shopifyclone.security.JwtService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final ArtisanProfileRepository artisanProfileRepository;
    private final MerchantProfileRepository merchantProfileRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.email())) {
            throw new IllegalArgumentException("Email already in use");
        }

        Role role = request.role() != null ? request.role() : Role.CUSTOMER;

        User user = User.builder()
                .firstName(request.firstName())
                .lastName(request.lastName())
                .email(request.email())
                .password(passwordEncoder.encode(request.password()))
                .role(role)
                .enabled(true)
                .build();

        userRepository.save(user);

        // Create associated profile based on role
        if (role == Role.ARTISAN) {
            ArtisanProfile profile = ArtisanProfile.builder()
                    .user(user)
                    .brandName(request.brandName() != null ? request.brandName() : user.getFirstName() + "'s Workshop")
                    .bio(request.bio())
                    .location(request.location())
                    .commissionRate(0.15)
                    .build();
            artisanProfileRepository.save(profile);
        } else if (role == Role.MERCHANT) {
            MerchantProfile profile = MerchantProfile.builder()
                    .user(user)
                    .brandName(request.brandName() != null ? request.brandName() : user.getFirstName() + "'s Brand")
                    .description(request.bio())
                    .build();
            merchantProfileRepository.save(profile);
        }

        String accessToken = jwtService.generateToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.of(accessToken, refreshToken,
                user.getId(), user.getEmail(),
                user.getFirstName(), user.getLastName(),
                user.getRole().name());
    }

    public AuthResponse login(LoginRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.email(), request.password())
        );

        User user = userRepository.findByEmail(request.email())
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        String accessToken = jwtService.generateToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.of(accessToken, refreshToken,
                user.getId(), user.getEmail(),
                user.getFirstName(), user.getLastName(),
                user.getRole().name());
    }
}
